### PR TITLE
Toolchain port for Sparc Leon3

### PR DESCRIPTION
**Pull request by lukileczo**
JIRA: RTOS-248

## Description

Create toolchain for Sparc Leon3 architecture
## Motivation and Context

JIRA: RTOS-248
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: Makefile.sparcv8leon3 [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+CRTEND := $(shell $(CC) $(CFLAGS) -print-file-name=crtend.o)
+PHOENIXLIB := $(shell $(CC) $(CFLAGS) -print-file-name=libphoenix.a)
+LDLIBS := $(PHOENIXLIB) $(GCCLIB) $(CRTBEGIN) $(CRTEND)
+
+OBJCOPY = $(CROSS)objcopy
+OBJDUMP = $(CROSS)objdump
+
+STRIP = $(CROSS)strip
+
+VADDR_KERNEL_INIT = 0x40000000
```
Comment (agkaminski) [inline Makefile.sparcv8leon3:36]: Why not standard kernel virtual address 0xc0000000?

Comment (lukileczo) [reply Makefile.sparcv8leon3:36]: Qemu for Leon3 doesn't emulate MMU, so I guess it has to be physical address.

**Review by agkaminski — APPROVED** [review]

**Comment by nalajcie** [comment]
is empty `scripts/mkimg-leon3.sh` needed in this PR?

**Comment by lukileczo** [comment]
Deleted `scripts/mkimg-leon3.sh` as it's not needed now

**Comment by agkaminski** [comment]
@lukileczo  This looks ready for review, please mark as such if it's ready indeed

File: toolchain/gcc-9.3.0-06-sparc-phoenix.patch
Comment (nalajcie) [inline toolchain/gcc-9.3.0-06-sparc-phoenix.patch:1]: - `06` patch number is now taken by `libstdcpp`
- we should add libstdcpp compilation for sparc also

File: Makefile.sparcv8leon3 [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+AR = $(CROSS)ar
+ARFLAGS = -r
+
+LD = $(CROSS)ld
+LDFLAGS := -z max-page-size=0x200 --gc-sections
+
+VADDR_KERNEL_INIT=40000000
+
+ifeq ($(KERNEL), 1)
+	STRIP = $(CROSS)strip
```
Comment (agkaminski) [inline Makefile.sparcv8leon3:32]: Add `-ffixed-g6` to kernel `CFLAGS` to prohibit kernel from destroying pic base register during syscall perhaps? Or is there some other mechanism that save content of this register?

File: toolchain/gcc-9.3.0-06-libstdcpp.patch [hunk 3]
```diff
@@ -49,6 +49,11 @@ diff -Naur gcc-9.3.0-orig/libstdc++-v3/configure gcc-9.3.0/libstdc++-v3/configur
 +            fi
 +            ;;
 +
++          sparc)
++            # TODO: remove -mflat when not needed
```
Comment (agkaminski) [inline toolchain/gcc-9.3.0-06-libstdcpp.patch:53]: I guess need for `-mflat` has been eliminated now

File: Makefile.sparcv8leon3 (hunk 2 above)
Comment (agkaminski) [reply Makefile.sparcv8leon3:32]: Not needed afterall

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**